geometry/util: fixed idiff() for Function support

verastrass opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #15584 

#### Brief description of what is fixed or changed
Now idiff(eq, y, x, n=1) supports:

- Eq as its first argument.
	Example:
		In [1]: from sympy import *
		In [2]: x, y = symbols('x y')
		In [3]: print(idiff(Eq(y*exp(y), x*exp(x)), y, x))
		Out [3]: (x + 1)*exp(x - y)/(y + 1)

- f(x) instead of y, where f is a Function.
	Example:
		In [1]: from sympy import *
		In [2]: x = symbols('x')
		In [3]: f = Function('f')
		In [4]: print(idiff(f(x)*exp(f(x)) - x*exp(x), f(x), x))
		Out [4]: (x + 1)*exp(x - f(x))/(f(x) + 1)

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* geometry
    * in idiff added support for the case when y is a Function
<!-- END RELEASE NOTES -->

verastrass committed:
geometry/util: Fixed issue 15584 with idiff

Now idiff(eq, y, x, n=1) supports:
(1) Eq as its first argument.
	Example:
		In [1]: from sympy import *
		In [2]: x, y = symbols('x y')
		In [3]: print(idiff(Eq(y*exp(y), x*exp(x)), y, x))
		Out [3]: (x + 1)*exp(x - y)/(y + 1)

(2) f(x) instead of y, where f is a Function.
	Example:
		In [1]: from sympy import *
		In [2]: x = symbols('x')
		In [3]: f = Function('f')
		In [4]: print(idiff(f(x)*exp(f(x)) - x*exp(x), f(x), x))
		Out [4]: (x + 1)*exp(x - f(x))/(f(x) + 1)

smichr commented on sympy/core/function.py:
```diff
@@ -1325,6 +1326,10 @@ def __new__(cls, expr, *variables, **kwargs):
         # _eval_derivative method
         if not evaluate or not hasattr(expr, '_eval_derivative'):
+            # return the derivative of Equality
+            if isinstance(expr, Equality):
```
We do not have an Equation class, just `Eq` and an `Eq` is a boolean. And although I think there is some code somewhere (present, or removed) that tried to implement this, differentiating "both sides" is incompatible with the current definition of "Eq".

smichr commented:
-1 (see comment). To utilize differentiation I believe it is the user's responsibility to unpack the Boolean into an Expr via `eq.rewrite(Add)`. While it makes sense for `solve` to see when the two sides are equal (since that would give the conditions for the equality being true) I don't think it makes sense to do so for the derivative which is supposed to tell how something -- in this case a Bool -- changes with respect to a variable.

verastrass committed:
geometry/util: changed the way of support for Eq in idiff()

If first argument is Eq then it's transformed to Expr.
Removed differentiation of Equality from __new__ method of Derivative.

smichr commented:
This looks good. Once a test is added this should be ready to go.

verastrass committed:
Added tests for idiff()

Added 5 tests in test_idiff() in geometry/tests/test_util.py

smichr commented on sympy/geometry/util.py:
```diff
@@ -570,12 +571,22 @@ def idiff(eq, y, x, n=1):
-        raise ValueError("expecting x-dependent symbol(s) but got: %s" % y)
+        raise ValueError("expecting x-dependent symbol(s) or function(s) but got: %s" % y)
+
+    if isinstance(eq, Equality):
```
oops! I missed this. A derivative of a bool is something that might be defined (see [here](https://github.com/sympy/sympy/issues/13350)) but if you use the `Eq` as an equation container in this routine, it will never be able to calculate a derivative of this boolean. If the *user* wants to use `Eq` as an equation container, that is their choice; if they want to use `idiff` *they* will have to pass it as `eq.rewrite(Add)`; this is not something that `idiff` should do.

smichr commented:
-1 for handling `Eq` in `idiff` routine (see comment).

smichr commented:
> is defined on an equation

But differentiation of a boolean is defined (though I do not understand the particulars) and `Eq` is a boolean. Until we have an `Eqn` class, I think it would be better not to introduce this unless it can be shown that `idiff` would never be expected to work with a boolean.

verastrass commented:
If we don't consider Eq in idiff as Expr, then we will have the same error given in #15584 using Eq. Is it okay? And maybe it's necessary to check if idiff's first argument is Expr and throw an exception otherwise?

smichr commented:
Please see the discussion [here](https://groups.google.com/forum/#!topic/sympy/WqQIp8oDbVk). If we make this change we will end up with a deprecation situation if 1) we do/can define differentiation of a boolean and 2) introduce an Eqn class.

verastrass committed:
removed support for Eq in idiff

Removed support for Eq in idiff and related tests.

smichr committed:
delete unused imports
